doc/kc.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "KC" "1" "January 2015" "" ""
.
.SH "NAME"
\fBkc\fR \- generic producer and consumer for Apache Kafka
.
.SH "SYNOPSIS"
.
.nf

kc producer [\-\-brokers=<brks>] [\-\-partition=<part>] [\-\-compression=<comp>]
            [\-\-delimiter=<delim>] [\-\-key\-delimiter=<delim>] [\-\-count=<cnt>]
            [\-\-error\-file=<file>] [\-T | \-\-tee] [\-q | \-\-quiet]
            [\-v | \-\-verbose] <topic> [<file>\|\.\|\.\|\.]

kc consumer [\-\-brokers=<brks>] [\-\-partition=<part>] [\-\-offset=<off>]
            [\-\-delimiter=<delim>] [\-\-key\-delimiter=<delim>] [\-\-count=<cnt>]
            [\-e | \-\-exit] [\-O | \-\-print\-offset] [\-q | \-\-quiet]
            [\-v | \-\-verbose] <topic>

kc metadata [\-\-brokers=<brks>] [\-\-partition=<part>][\-q | \-\-quiet]
            [\-v | \-\-verbose] [<topic>]

kc \-\-help

kc \-\-version
.
.fi
.
.SH "DESCRIPTION"
\fBkc\fR also known as \fBkafkacat\fR is a generic non\-JVM producer and consumer for Apache Kafka 0\.8, think of it as a netcat for Kafka\.
.
.P
\fBkc\fR works in different \fIcommand\fR: producer, consumer and metadata\. Each one is described in the [COMMAND] section\.
.
.SH "COMMAND"
.
.IP "\[ci]" 4
\fBproducer\fR reads messages from stdin, delimited with a configurable delimeter and produces them to the provided Kafka cluster, topic and partition\.
.
.IP "\[ci]" 4
\fBconsumer\fR reads messages from a topic (and possibly a specific partition ) and prints them to stdout using the configured message delimiter\.
.
.IP "\[ci]" 4
\fBmetadata\fR displays the current state of the Kafka cluster and its topics and partitions\.
.
.IP "" 0
.
.SH "OPTIONS"
.
.SS "Generic options"
.
.IP "\[ci]" 4
\fB\-b <brokers>, \-\-brokers=<brokers>\fR Comma separated list of broker(s) to bootstrap of the form "host[:port][,\|\.\|\.\|\.]", e\.g\. "host1:9292,host2:9293" [default: localhost]\.
.
.IP "\[ci]" 4
\fB\-p <partition>, \-\-partition=<partition>\fR Send messages to a specific partition\. If \-1 is provided, a partition will be randomly uniformly selected for each message\.
.
.IP "\[ci]" 4
\fB\-v, \-\-verbose\fR Augment verbosity level\.
.
.IP "\[ci]" 4
\fB\-q, \-\-quiet\fR Quiet mode, do not report error messages to stderr\.
.
.IP "\[ci]" 4
\fB\-V, \-\-version\fR Show the version on the first line and follow with the usage and help message\.
.
.IP "\[ci]" 4
\fB\-h, \-\-help\fR Show the usage and help message\.
.
.IP "" 0
.
.SS "Producer options"
.
.IP "\[ci]" 4
\fB\-d <delim>, \-\-delimiter=<delim>\fR Message delimiter character: a\-z\.\. | \er | \en | \et | \exNN [default: \en]\.
.
.IP "\[ci]" 4
\fB\-k <delim>, \-\-key\-delimiter=<delim>\fR Key delimiter character: a\-z\.\. | \er | \en | \et | \exNN\.
.
.IP "\[ci]" 4
\fB\-c <cnt>, \-\-count=<cnt>\fR Exit after producing \fB<cnt>\fR messages\.
.
.IP "\[ci]" 4
\fB\-z <c>, \-\-compression=<c>\fR Compression applied to messages: none, snappy or gzip [default: none]\.
.
.IP "\[ci]" 4
\fB\-T, \-\-tee\fR Output queued messages to stdout, acting like tee\. A message is relayed to stdout once it hits librdkafka\'s internal queue, it can still fail to reach the producer\. See \fB\-E\fR for more information on failed messages\.
.
.IP "\[ci]" 4
\fB\-E <file>, \-\-error\-file=<file>\fR Messages that couldn\'t be sent are append to \fB<file>\fR, formatted according to delimiter and key delimiter (if provided)\. There is not guarantee on the ordering of the messages [default: stderr]\.
.
.IP "" 0
.
.SS "Consumer options"
.
.IP "\[ci]" 4
\fB\-d <delim>, \-\-delimiter=<delim>\fR Message delimiter character: a\-z\.\. | \er | \en | \et | \exNN [default: \en]\.
.
.IP "\[ci]" 4
\fB\-k <delim>, \-\-key\-delimiter=<delim>\fR Key delimiter character: a\-z\.\. | \er | \en | \et | \exNN\.
.
.IP "\[ci]" 4
\fB\-o <off>, \-\-offset=<off>\fR Define the consuming offset, the possible values are:
.
.IP "\[ci]" 4
\fBbeginning\fR: start from the beginning
.
.IP "\[ci]" 4
\fBend\fR: start from the end
.
.IP "\[ci]" 4
\fBstored\fR: start from a stored value
.
.IP "\[ci]" 4
\fB<value>\fR: start at absolute \fIvalue\fR
.
.IP "\[ci]" 4
\fB\-<value>\fR: start at relative offset from the end (acting like \fBtail\fR)\.
.
.IP "" 0

.
.IP "\[ci]" 4
\fB\-c <cnt>, \-\-count=<cnt>\fR Exit after consuming \fB<cnt>\fR messages\.
.
.IP "\[ci]" 4
\fB\-e , \-\-exit\fR Exit when there is no message left to consumer in the topic\.
.
.IP "" 0
.
.SH "EXAMPLES"
.
.IP "\[ci]" 4
Producing a single message
.
.IP "" 0
.
.IP "" 4
.
.nf

    $ echo "test message" | kc producer test
.
.fi
.
.IP "" 0
.
.IP "\[ci]" 4
Consuming last message
.
.IP "" 0
.
.IP "" 4
.
.nf

    $ kc consumer \-e \-o \-1 test
    test message
.
.fi
.
.IP "" 0
.
.IP "\[ci]" 4
Showing metadata on the test topic
.
.IP "" 0
.
.IP "" 4
.
.nf

    $ kc metadata test
    {
      "brokers": [
        { "id":0, "host":"localhost:9092" }
      ],
      "topics": [
        {
          "name": "test",
          "partitions": [
            { "id": 0, "leader": 0, "replicas": [0], "isrs": [0] }
          ]
        }
      ]
    }
.
.fi
.
.IP "" 0



src/kc.c
/*
 * kc - Apache Kafka consumer and producer
 *
 * Copyright (c) 2015, François Saint-Jacques
 * Copyright (c) 2014, Magnus Edenhill
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "common.h"
#include "../config.h"

typedef enum {
  UNKNOWN,
  PRODUCER,
  CONSUMER,
  METADATA,
} kc_command;

int producer_main(int argc, char **argv);
int consumer_main(int argc, char **argv);
int metadata_main(int argc, char **argv);

static struct option kc_long_options[] = {
    {"help",    no_argument, 0, 'h'},
    {"version", no_argument, 0, 'V'},
    {0,         0,           0,  0 }
};

static kc_command argparse (int argc, char **argv) {
  int opt;
  int option_index = 0;

  /* Prefixing the optstring with '+' will have the effect of
   * stopping getopt_long on the first non option. In our case, this
   * would be the command. */
  while ((opt = getopt_long(argc, argv,
                            "+Vh",
                            kc_long_options,
                            &option_index)) != -1) {
    switch (opt) {
    case 'V':
      usage(argv[0], 0, KC_VERSION_STR);
      break;
    case 'h':
      usage(argv[0], 0, NULL);
      break;
    default:
      usage(argv[0], 1, NULL);
      break;
    }
  }

  if (argc - optind == 0)
    usage(argv[0], 1, "Command missing");

  const char *cmd = argv[optind];
  if (!strcmp("producer", cmd))
    return PRODUCER;
  else if (!strcmp("consumer", cmd))
    return CONSUMER;
  else if (!strcmp("metadata", cmd))
    return METADATA;

  return UNKNOWN;
}

int main(int argc, char **argv) {
  int    left_argc;
  char **left_argv;

  set_signals();

  /* Create config containers */
  conf.rk_conf  = rd_kafka_conf_new();
  conf.rkt_conf = rd_kafka_topic_conf_new();

  kc_command cmd = argparse(argc, argv);

  left_argc = argc - optind;
  left_argv = &argv[optind];

  /* Execute command */
  switch (cmd) {
  case PRODUCER:
    producer_main(left_argc, left_argv);
    break;
  case CONSUMER:
    consumer_main(left_argc, left_argv);
    break;
  case METADATA:
    metadata_main(left_argc, left_argv);
    break;
  default:
    usage(argv[0], 1, "Unknown subcommand");
  }

  /* Be warned that changing conf.run is highly risky. */
  conf.run = 1;

  /* Wait for all messages to be transmitted */
  while (conf.run && rd_kafka_outq_len(conf.rk))
    rd_kafka_poll(conf.rk, 50);

  if (conf.rkt) {
    rd_kafka_topic_destroy(conf.rkt);
    conf.rkt = NULL;
  }

  if (conf.rk) {
    rd_kafka_destroy(conf.rk);
    conf.rk = NULL;
  }

  rd_kafka_wait_destroyed(5000);

  return conf.exitcode;
}
